api/src/main/java/org/openmrs/CohortMembership.java

```diff
@@ -99,6 +118,18 @@ public void setEndDate(Date endDate) {
 
 	@Override
 	public int compareTo(CohortMembership o) {
-		return this.getPatient().getPatientId() - o.getPatient().getPatientId();
+		if (Objects.equals(this.getPatient().getPatientId(), o.getPatient().getPatientId())
+				&& Objects.equals(this.getCohort().getCohortId(), o.getCohort().getCohortId())
+				&& this.getStartDate().equals(o.getStartDate())
+				&& ((this.getEndDate() != null && o.getEndDate() != null
+				&& this.getEndDate().equals(o.getEndDate()))
+				|| (this.getEndDate() == null && o.getEndDate() == null))) {
+			return 0;
+		} else if ((this.isMembershipActive() && !o.isMembershipActive())
+			|| (!this.isMembershipActive() && o.isMembershipActive())) {
```

Comment:
If this membership is active and the other is inactive, you should return -1, then return 1 for the opposite, in your code above you are returning -1 both cases which I think is incorrect.

The other thing is that the first condition isn't sorting by startDates since you always return 0 which means they are equal.